Add API-M Tutorial scenarios (#3715)
wso2/docs-apim: en/docs/tutorials/scenarios/scenario14-external-key-manager.md

# Scenario 14 - External Key Manager Support

## User Story

> **RailCo has users in their internal identity management system. They do not want to have duplicate user information in two systems. RailCo wants to use this IDP to generate access tokens to consume their APIs without having to import their users to the GOGO train API platform.**

Reviewer: ```suggestion
RailCo has users in their internal identity management system. They do not want to have duplicate user information in two systems. RailCo wants to use this IDP to generate access tokens to consume their APIs without having to import their users to the GOGO train API platform.
```

Author: Fixed

**_Time to Complete : 5 mins_**

> WSO2 API Manager comes with out-of-the-box external key manager connectors to various Identity providers such as Okta, Auth0, Keycloak, etc. Additionally it has provided interfaces to write custom keymanager implementation to plug any external third party keymanager solution. 

Reviewer: ```suggestion
WSO2 API Manager comes with out-of-the-box external key manager connectors to various Identity providers such as Okta, Auth0, Keycloak, etc. Additionally, it has provided interfaces to write custom Key Manager implementation to plug any external third-party Key Manager solution. 
```

Author: Fixed

For this scenario, we assume Railco has Keycloak IDP as their internal identity provider. Following are the steps to configure Keycloak idp as the external keymanager for RailCo.

![KM description]({{base_path}}/assets/img/tutorials/scenarios/km-desc.png)

> Detailed steps on how to configure Keycloak can be found in [WSO2 Documentation](https://apim.docs.wso2.com/en/latest/administer/key-managers/configure-keycloak-connector/) . Additionally, you could refer to the screen cast done on this in [here](https://www.youtube.com/watch?v=xuZ6DPhXNX8). Following are the simplified steps for this setup.

Reviewer: ```suggestion
Detailed steps on how to configure Keycloak can be found in [WSO2 Documentation]({{base_path}}/administer/key-managers/configure-keycloak-connector/). Additionally, you could refer to the screencast done on this in [here](https://www.youtube.com/watch?v=xuZ6DPhXNX8). Following are the simplified steps for this setup.
```

Author: Fixed

1. Download Keycloak Server from [https://www.keycloak.org/downloads](https://www.keycloak.org/downloads). At the time of writing this tutorial, keycloak 12.0.4 was used for this.
2. Extract it to your local setup and navigate to `<keycloak>/bin` and run `./standalone.sh` to start the server.

> 3. Go To [http://localhost:8080/](http://localhost:8080/) . If this is the first time, create an initial admin user from the form given in this page.

Reviewer: ```suggestion
3. Go To [http://localhost:8080/](http://localhost:8080/). If this is the first time, create an initial admin user from the form given in this page.
```

Author: Fixed

4. Run the following command. Replace admin username and password. This will run a script on keycloak to generate application on keycloak.

```
docker run -e USER=<admin_username> -e PASSWORD=<admin_password> chamilaadhi/keycloak-apim-script:1.0.0
```

If execution is successful, you will be given a client Id and client secret to use with WSO2 API Manager

```

=====================================================

client id    :  apim-client
client secret:  221f10b7-d169-45e0-851e-e1b017052162

=====================================================

```

!!!Note

>     Script used in this docker container to create the credentials can be found in `/resource/setup-keycloak.sh` file

Reviewer: ```suggestion
    The script used in this Docker container to create the credentials can be found in the `/resource/setup-keycloak.sh` file.
```

Author: Fixed

Now let's configure the connection between API-M and Keycloak.

> 1. Login to admin portal [https://localhost:9443/admin/](https://localhost:9443/admin/) using railco admin user. Use admin@railco.com and _admin_ as the password.

Reviewer: ```suggestion
1. Log on to admin portal [https://localhost:9443/admin/](https://localhost:9443/admin/) using RailCo admin user. Use `admin@railco.com` and `_admin_` as the password.
```

Author: Fixed
